crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -0,0 +1,144 @@
+                self.base_url, self.chain
+            );
+            let mut builder = self.client.get(&url);
+            if let Some(ref api_key) = self.api_key {
+                builder = builder.header(AUTHORIZATION, api_key)
+            }
+            observe::coingecko_request(&url);
+            let response = builder.send().await;
+            observe::coingecko_response(&url, response.as_ref());
+            let response = response.map_err(|e| {
+                PriceEstimationError::EstimatorInternal(anyhow!(
+                    "failed to sent CoinGecko price request: {e:?}"
+                ))
+            })?;
+            if !response.status().is_success() {
+                let status = response.status();
+                return match status {
+                    StatusCode::TOO_MANY_REQUESTS => Err(PriceEstimationError::RateLimited),
+                    status => Err(PriceEstimationError::EstimatorInternal(anyhow!(
+                        "failed to retrieve prices from CoinGecko: error with status code \
+                         {status}."
+                    ))),
+                };
+            }
+            let response = response.text().await.map_err(|e| {
+                PriceEstimationError::EstimatorInternal(anyhow!(
+                    "failed to fetch native CoinGecko prices: {e:?}"
+                ))
+            })?;
+            let prices = serde_json::from_str::<Response>(&response)
+                .map_err(|e| {
+                    PriceEstimationError::EstimatorInternal(anyhow!(
+                        "failed to parse native CoinGecko prices from {response:?}: {e:?}"
+                    ))
+                })?
+                .0;
+
+            let price = prices
+                .get(&token)
+                .ok_or(PriceEstimationError::NoLiquidity)?;
+            Ok(price.eth)
+        }
+        .boxed()
+    }
+}
+
+mod observe {
+    use reqwest::Response;
+
+    /// Observe a request to be sent to CoinGecko
+    pub fn coingecko_request(endpoint: &str) {
+        tracing::trace!(%endpoint, "sending request to CoinGecko");
+    }
+
+    /// Observe that a response was received from CoinGecko
+    pub fn coingecko_response(endpoint: &str, res: Result<&Response, &reqwest::Error>) {
+        match res {
+            Ok(res) => {
+                tracing::trace!(%endpoint, ?res, "received response from CoinGecko")
+            }
+            Err(err) => {
+                tracing::warn!(%endpoint, ?err, "failed to receive response from CoinGecko")
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {super::*, std::str::FromStr};
+
+    // It is ok to call this API without an API for local testing purposes as it is
+    // difficulty to hit the rate limit manually
+    const BASE_URL: &str = "https://api.coingecko.com/api/v3/simple/token_price";
+
+    #[tokio::test]
+    #[ignore]
+    async fn works() {
```

Comment:
Should this be executed on CI? I believe it is ignored currently since we don't have a separate job to run ignored shared tests.

Reply:
I thought of that, but then depending on a third party in order to pass a test is not a good idea. What's more, we could theoretically hit the rate limit (it is difficult but possible), or they can change the rate limit policies too, etc.

I would keep it as `ignore` the same as the tests for `1inch` for example.